### PR TITLE
New Resource: `azurerm_sql_managed_instance_active_directory_administrator`

### LINKED ISSUE #13957: Support for Active Directory admin  of a SQL Managed Instance
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

When using Azure Portal, you can create a system managed identity/Service Principle and then set an Active Directory Admin for the instance.  You can optionally set if you want to Support only Azure Active Directory authentication for this managed instance
image

this could be enabled in the same way as azurerm_sql_active_directory_administrator

### New or Affected Resource(s)

* azurerm_sql_managed_instance_administrator

### Potential Terraform Configuration

```hcl
resource "azurerm_sql_managed_instance_administrator" "example" {
  server_name            = azurerm_sql_managed_instance.example.name
  resource_group_name    = azurerm_resource_group.example.name
  tenant_id              = data.azurerm_client_config.current.tenant_id
  object_id              = data.azuread_group.example_AD_Admin.object_id
  AD_auth_only           = false 
}

```

### PR DESCRIPTION
**Pull request by aristosvo**
Fixes #13957

Needs `azuread` v2 via #14105
Needs merge of #14052

## Acceptance Tests
```bash
TF_ACC=1 go test -v ./internal/services/sql -run=TestAccSqlMiAdministrator_basic -timeout 1800m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccSqlMiAdministrator_basic
=== PAUSE TestAccSqlMiAdministrator_basic
=== CONT  TestAccSqlMiAdministrator_basic
--- PASS: TestAccSqlMiAdministrator_basic (17391.73s)
PASS
ok  	github.com/hashicorp/terraform-provider-azurerm/internal/services/sql	17391.752s
```

### CODE REVIEW

**Comment by katbyte** [comment]
Hmm, looks like we need to increase the timeouts even more?
```
------- Stdout: -------
=== RUN   TestAccAzureRMSqlMiServer_basic
=== PAUSE TestAccAzureRMSqlMiServer_basic
=== CONT  TestAccAzureRMSqlMiServer_basic
    testcase.go:110: Step 1/2 error: Error running apply: exit status 1
        
        Error: sql.ManagedInstancesClient#CreateOrUpdate: Failure sending request: StatusCode=504 -- Original Error: Code="GatewayTimeout" Message="The gateway did not receive a response from 'Microsoft.Sql' within the specified time period."
        
          with azurerm_sql_managed_instance.test,
          on terraform_plugin_test.tf line 1142, in resource "azurerm_sql_managed_instance" "test":
        1142: resource "azurerm_sql_managed_instance" "test" {
        
--- FAIL: TestAccAzureRMSqlMiServer_basic (1289.87s)
FAIL
```

**Comment by aristosvo** [comment]
@katbyte which timeouts can we possibly change here? This seems like a timeout of the Azure API, not a resource creation timeout AFAIK.

The resource create timeout for a Managed Instance is 24hrs, which is much more than current the 20 minutes timeout we're facing..?

**Comment by aristosvo** [comment]
Currently waiting for response on https://github.com/Azure/azure-rest-api-specs/issues/16838

File: internal/services/sql/sql_managed_instance_administrator_resource.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, id.ResourceGroup, id.ManagedInstanceName)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for presence of existing %q: %+v", id, err)
+			}
+		}
+
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_sql_active_directory_administrator", *existing.ID)
```
Comment (aristosvo) [inline internal/services/sql/sql_managed_instance_administrator_resource.go:95]: Oops
```suggestion
			return tf.ImportAsExistsError("azurerm_sql_managed_instance_active_directory_administrator", *existing.ID)
```

File: internal/services/sql/sql_managed_instance_administrator_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+func (r SqlMiAdministratorResource) requiresImport(data acceptance.TestData) string {
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_sql_managed_instance_active_directory_administrator" "import" {
+  server_name         = azurerm_sql_active_directory_administrator.test.server_name
+  resource_group_name = azurerm_sql_active_directory_administrator.test.resource_group_name
+  login               = azurerm_sql_active_directory_administrator.test.login
+  tenant_id           = azurerm_sql_active_directory_administrator.test.tenant_id
+  object_id           = azurerm_sql_active_directory_administrator.test.object_id
```
Comment (aristosvo) [inline internal/services/sql/sql_managed_instance_administrator_resource_test.go:127]:
```suggestion
resource "azurerm_sql_managed_instance_active_directory_administrator" "import" {
  server_name         = azurerm_sql_managed_instance_active_directory_administrator.test.server_name
  resource_group_name = azurerm_sql_managed_instance_active_directory_administrator.test.resource_group_name
  login               = azurerm_sql_managed_instance_active_directory_administrator.login
  tenant_id           = azurerm_sql_managed_instance_active_directory_administrator.test.tenant_id
  object_id           = azurerm_sql_managed_instance_active_directory_administrator.test.object_id
```

**Review by aristosvo — COMMENTED** [review]
Another error on import

**Comment by aristosvo** [comment]
And last test fixed:
```bash
--- PASS: TestAccSqlMiAdministrator_requiresImport (15679.28s)
```

File: internal/services/sql/sql_managed_instance_administrator_resource.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+			"tenant_id": {
+				Type:         pluginsdk.TypeString,
+				Required:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+
+			"azuread_authentication_only": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Computed: true,
```
Comment (mbfrahry) [inline internal/services/sql/sql_managed_instance_administrator_resource.go:68]: It looks like this is defaulted to false when we create it in the portal? Should we follow that same logic here?

File: website/docs/r/sql_managed_instance_active_directory_administrator.html.markdown [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+* `managed_instance_name` - (Required) The name of the SQL Managed Instance on which to set the administrator. Changing this forces a new resource to be created.
+
+* `resource_group_name` - (Required) The name of the resource group for the SQL Managed Instance. Changing this forces a new resource to be created.
+
+* `login` - (Required) The login name of the principal to set as the Managed Instance administrator
+
+* `object_id` - (Required) The ID of the principal to set as the Managed Instance administrator
+
+* `tenant_id` - (Required) The Azure Tenant ID
+
```
Comment (mbfrahry) [inline website/docs/r/sql_managed_instance_active_directory_administrator.html.markdown:58]: `azuread_authentication_only` is missing from the docs

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @aristosvo! This PR looks good with just a couple callouts

**Comment by aristosvo** [comment]
Rebased and squashed!

File: internal/services/sql/sql_managed_instance_administrator_resource.go (hunk 3 above)
Comment (aristosvo) [reply internal/services/sql/sql_managed_instance_administrator_resource.go:68]: This is fine by me. It's better than `Computed`, question is then whether we should make the call to set it explicitly (even if `azuread_authentication_only = false`). I implemented it now in that way, that's the most secure way, but can be unnecessary.

**Review by mbfrahry — APPROVED** [review]
LGTM! Thanks @aristosvo

**Pull request merged**